src/search/task_proxy.h

```diff
@@ -729,13 +718,90 @@ inline bool does_fire(const EffectProxy &effect, const State &state) {
     return true;
 }
 
-inline bool does_fire(const EffectProxy &effect, const GlobalState &state) {
-    for (FactProxy condition : effect.get_conditions()) {
-        FactPair condition_pair = condition.get_pair();
-        if (state[condition_pair.var] != condition_pair.value)
-            return false;
+inline bool State::operator==(const State &other) const {
+    assert(task == other.task);
+    if (registry != other.registry) {
+        std::cerr << "Comparing registered states with unregistered states "
+                  << "or registered states from different registries is "
+                  << "treated as an error because it is likely not "
+                  << "intentional."
+                  << std::endl;
+        utils::exit_with(utils::ExitCode::SEARCH_CRITICAL_ERROR);
+    }
+    if (registry) {
+        // Both states are registered and from the same registry.
+        return id == other.id;
+    }
+    // Both states are unregistered.
+    assert(values);
+    assert(other.values);
+    assert(!values->empty());
+    return *values == *other.values;
+}
+
+inline bool State::operator!=(const State &other) const {
+    return !(*this == other);
+}
+
+inline void State::unpack() const {
+    if (!values) {
+        int num_variables = size();
+        values = std::make_shared<std::vector<int>>(num_variables);
+        for (int var = 0; var < num_variables; ++var) {
+            (*values)[var] = state_packer->get(buffer, var);
+        }
+    }
+}
+
+inline std::size_t State::size() const {
+    return num_variables;
+}
+
+inline FactProxy State::operator[](std::size_t var_id) const {
+    assert(var_id < size());
+    if (values) {
+        return FactProxy(*task, var_id, (*values)[var_id]);
+    } else {
+        assert(buffer);
+        assert(state_packer);
+        return FactProxy(*task, var_id, state_packer->get(buffer, var_id));
     }
-    return true;
 }
 
+inline FactProxy State::operator[](VariableProxy var) const {
+    return (*this)[var.get_id()];
+}
+
+inline TaskProxy State::get_task() const {
+    return TaskProxy(*task);
+}
+
+inline const StateRegistry *State::get_registry() const {
+    return registry;
+}
+
+inline StateID State::get_id() const {
+    return id;
+}
+
+inline const PackedStateBin *State::get_buffer() const {
+    if (!buffer) {
+        std::cerr << "Accessing the packed values of an unregistered state is "
+                  << "treated as an error."
+                  << std::endl;
+        utils::exit_with(utils::ExitCode::SEARCH_CRITICAL_ERROR);
+    }
+    return buffer;
+}
+
+inline const std::vector<int> &State::get_unpacked_values() const {
```

Comment:
I thought using `inline` was frowned upon because IIRC compilers ignore these directives anyway.

Reply:
In this place (the definition outside the class) it is needed to avoid linker errors, it has the effect of enforcing static linkage. It is frowned about for the declaration.